examples/nlp/language_modeling/megatron_change_num_partitions.py

```diff
@@ -830,98 +909,215 @@ def main():
+            if vp_size > 1:
+                set_virtual_parallel_rank_safely(vp_rank)
 
-                model = cls.restore_from(
-                    restore_path=model_filepath,
-                    trainer=trainer,
-                    map_location=torch.device("cpu"),
-                    save_restore_connector=save_restore_connector,
-                )
-                model.to(dtype=dtype)
+            for pp_rank in range(pp_size):
+                app_state.pipeline_model_parallel_rank = pp_rank
 
-                # Reset env flag
-                os.environ.pop(NEMO_MEGATRON_MODEL_PARALLEL_APPSTATE_OVERRIDE, None)
+                for tp_rank in range(tp_size):
+                    app_state.tensor_model_parallel_rank = tp_rank
 
-                logging.info(
-                    f"<<<<<<<< LOADED MODEL PP={pp_rank + 1} TP={tp_rank + 1} | "
-                    f"GLOBAL RANK = {global_rank} >>>>>>>>>"
-                )
-                params = [p for _, p in model.named_parameters()]
-                partitions[pp_rank].append(params)
+                    logging.info(f"Loading ------------ PP Rank: {pp_rank} TP Rank: {tp_rank}")
 
-                # app_state is being updated incorrectly during restore
-                app_state.data_parallel_rank = 0
-                app_state.pipeline_model_parallel_rank = pp_rank
-                app_state.tensor_model_parallel_rank = tp_rank
-                app_state.pipeline_model_parallel_size = pp_size
-                app_state.tensor_model_parallel_size = tp_size
-                app_state.model_parallel_size = (
-                    app_state.pipeline_model_parallel_size * app_state.tensor_model_parallel_size
-                )
+                    # Override flag that forces Model to use AppState instead of Trainer
+                    # to determine the world size, global and local rank
+                    # Used for simulating load of a specific rank on a single gpu
+                    os.environ[NEMO_MEGATRON_MODEL_PARALLEL_APPSTATE_OVERRIDE] = "true"
+
+                    # Compute the global rank to load the correct subset of parameters
+                    global_rank = pp_rank * tp_size + tp_rank
+
+                    # Update AppState
+                    app_state.world_size = world_size
+                    app_state.global_rank = global_rank
+                    app_state.local_rank = global_rank % num_gpu_per_node
+                    app_state.pipeline_model_parallel_size = pp_size
+                    app_state.tensor_model_parallel_size = tp_size
+                    app_state.pipeline_model_parallel_split_rank = pipeline_model_parallel_split_rank
+                    app_state.model_parallel_size = (
+                        app_state.pipeline_model_parallel_size * app_state.tensor_model_parallel_size
+                    )
+
+                    if vp_size > 1:
+                        set_virtual_parallel_rank_safely(vp_rank)
```

Comment:
why are we doing this again? we seem to have already set it the outside loop

Reply:
This gets reset to different values in parts of the script so for safety i reset this back to the value outside the loop. It can possibly be removed, but better keep it.